### PR TITLE
Populate fingerprint properties automatically

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Resolve https://github.com/pulumi/pulumi-google-native/issues/126

1. Detect fingerprint properties during code generation and save them to metadata. Unfortunately, I had to opt-in several modules to this model and leave everything else out because other modules contain properties called "fingerprint" or similar but those properties need to be populated manually by users, e.g. to provide a hash of a file or a certificate ([example](https://github.com/pulumi/pulumi-google-native/blob/0471720df02f1ad3071f5fcc223ce1a617e45c7d/discovery/firebaserules_v1.json#L537)).
2. The provider's Update uses the old state to copy fingerprint values according to metadata.
3. Don't add fingerprint properties to Pulumi schema and SDKs.

### CODE REVIEW

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -359,11 +359,13 @@ func (p *googleCloudProvider) Create(ctx context.Context, req *rpc.CreateRequest
 	}, nil
 }
 
+
 func (p *googleCloudProvider) prepareAPIInputs(
-	inputs resource.PropertyMap,
+	inputs resource.PropertyMap, state resource.PropertyMap,
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:364]: nit: You can declare args as a comma-separated list if they are the same type, e.g.
```go
inputs, state resource.PropertyMap,
```

File: provider/pkg/resources/convert.go [hunk 2]
```diff
@@ -15,9 +15,9 @@ type SdkShapeConverter struct {
 	Types map[string]CloudAPIType
 }
 
-type convertPropValues func(props map[string]CloudAPIProperty, values map[string]interface{}) map[string]interface{}
+type convertPropValues func(props map[string]CloudAPIProperty, inputs map[string]interface{}, state map[string]interface{}) map[string]interface{}
```
Comment (lblackstone) [inline provider/pkg/resources/convert.go:18]: Same here

**Review by lblackstone — APPROVED** [review]

File: provider/pkg/gen/schema.go [hunk 3]
```diff
@@ -718,6 +723,19 @@ func (g *packageGenerator) genProperties(typeName string, typeSchema *discovery.
 	return &result, nil
 }
 
+func (g *packageGenerator) isCopyFromOutput(propName string) bool {
```
Comment (viveklak) [inline provider/pkg/gen/schema.go:726]: Nit - rename to `shouldCopyFromOutput`?  Current name suggests something was copied already.

File: provider/pkg/gen/schema.go [hunk 4]
```diff
@@ -718,6 +723,19 @@ func (g *packageGenerator) genProperties(typeName string, typeSchema *discovery.
...
+	mod := strings.Split(g.mod, "/")[0]
+	switch mod {
+	case "compute", "container", "deploymentmanager":
+		return strings.Contains(strings.ToLower(propName), "fingerprint")
+	default:
+		// Some other modules contain properties called "fingerprint" or similar but that need to be
+		// populated manually by users, e.g. to provide a hash of a file or a certificate. Therefore,
+		// we maintain the opt-in list of modules above and treat the rest as normal inputs.
+		return false
```
Comment (viveklak) [inline provider/pkg/gen/schema.go:735]: I am curious how we want to go about detecting that our opt-in/out lists are stale? Does it make sense to perhaps generate explicit lists in code and flag any deviation when we refresh discovery docs etc? Happy to track a ticket to track this if you like.

**Review by viveklak — APPROVED** [review]

File: provider/pkg/gen/schema.go (hunk 4 above)
Comment (mikhailshilkov) [reply provider/pkg/gen/schema.go:735]: Yes, that's a valid concern and a good idea to cut the ticket 👍

Comment (viveklak) [reply provider/pkg/gen/schema.go:735]: Opened https://github.com/pulumi/pulumi-google-native/issues/133 to track this.

**Pull request merged**